src/net/server/handlers/message_stream_layer/service.rs

```diff
@@ -0,0 +1,134 @@
+    ResBody: Default,
+    T: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match &mut self.sender {
+            SenderStatus::NotReady(sender) => {
+                if ready!(sender.poll_reserve(cx)).is_ok() {
+                    // swap `NotReady` for `Ready`
+                    if let SenderStatus::NotReady(sender) =
+                        std::mem::replace(&mut self.sender, SenderStatus::Spent)
+                    {
+                        self.sender = SenderStatus::Ready(sender);
+                        self.inner.poll_ready(cx)
+                    } else {
+                        panic!("SenderStatus was not NotReady")
+                    }
+                } else {
+                    // if `poll_reserve` returns error, forward error to response
+                    self.sender = SenderStatus::Err;
+                    Poll::Ready(Ok(()))
+                }
+            }
+            SenderStatus::Ready(_) => self.inner.poll_ready(cx),
+            SenderStatus::Err | SenderStatus::Spent => Poll::Ready(Ok(())),
+        }
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        match &self.sender {
+            SenderStatus::NotReady(_) => panic!("poll_ready not called"),
+            SenderStatus::Ready(_) => {
+                if let SenderStatus::Ready(sender) =
+                    std::mem::replace(&mut self.sender, SenderStatus::Spent)
+                {
+                    // must insert value wrapped in `Some` due to `Extension`'s `FromRequest` impl
+                    // calling `.clone()` on all `Extension`s. Thus, we must call `.extensions_mut` on
+                    // the `Request` inside the handler, which does NOT call `.clone()`, but which means
+                    // we must `.take()` the `sender` out of the `Extension` to prevent borrow contention
+                    // on the `Request`.
+                    req.extensions_mut().insert(Some(ReservedPermit(sender)));
+                    ResponseFuture::Inner(self.inner.call(req))
+                } else {
+                    panic!("SenderStatus was not Ready")
+                }
+            }
+            SenderStatus::Err => ResponseFuture::PollErr,
+            SenderStatus::Spent => panic!("call already called"),
+        }
+    }
+}
+
+/// When cloning, the `PollSender` does not transfer any permits it has acquired.
+/// Thus, `clone` is equivalent to initializing a new [`UpstreamMiddleware`] with the same `sender`
+impl<S: Clone, T> Clone for MessageStream<S, T> {
+    // manually implement Clone to get around requirements that `T` must be [`Clone`]
+    fn clone(&self) -> Self {
+        // Enforces that `.clone()` is never called after `.poll_ready()` has been called
+        #[cfg(debug_assertions)]
```

Comment:
Don't we need this check everywhere? it is cheap

Reply:
sure. removed the cfg